seisbench/models/dkpn.py

```diff
@@ -0,0 +1,784 @@
+
+        # --- NORMALIZE  input stream matrix (3C) with the matrix standard deviation
+        MM = MM / (
+            np.std(MM, axis=1, keepdims=True) + self.eps
+            )
+
+        # ===========================  GO CALCULATE
+        fp_cf_waveforms = []
+        fp_band_data = []
+        fp_max_arrays_band = []
+        fp_max_arrays_val = []
+
+        for _idx, arr in enumerate(MM):
+
+            # ====================
+            # 1) Run FBsummary
+            summary = FBSummary(arr,
+                                sampling_rate=100.0,
+                                npts=arr.shape[0],
+                                t_long=self.t_long,
+                                freqmin=self.freqmin,
+                                corner=self.corner,
+                                perc_taper=self.perc_taper,
+                                mode=self.mode)
+
+            fp_cf_waveforms.append(summary.summary)
+
+            fp_band_data.append(summary.BF)
+            fp_max_array_band = np.argmax(summary.FC, axis=0)
+            fp_max_arrays_band.append(fp_max_array_band)
+            fp_max_array_val = summary.summary
+            fp_max_arrays_val.append(fp_max_array_val)
+
+        fp_cf_waveforms = np.asarray(fp_cf_waveforms)
+
+        if self.clip > 0.0:
+            fp_cf_waveforms = np.clip(fp_cf_waveforms, a_min=None, a_max=self.clip)
+        if self.log:
+            fp_cf_waveforms = np.log10(fp_cf_waveforms + 1.0)
+        if self.normalize:
+            fp_cf_waveforms = fp_cf_waveforms / (
+                np.amax(np.abs(fp_cf_waveforms), axis=(0, 1), keepdims=True) + self.eps)
+
+        # ====================
+        # 2) Run polarization
+        fp_band_data = np.asarray(fp_band_data)
+        fp_max_arrays_band = np.asarray(fp_max_arrays_band)     # [nchan, npts]
+        fp_max_arrays_band_amax = np.amax(fp_max_arrays_band, axis=0)       # [npts]
+        band_amax = -1
+        if self.use_amax_only:
+            fp_max_arrays_val = np.asarray(fp_max_arrays_val)       # [nchan, npts]
+            fp_max_arrays_val_argmax_ndx = np.argmax(fp_max_arrays_val)
+            index_amax = np.unravel_index(fp_max_arrays_val_argmax_ndx, fp_max_arrays_val.shape)
+            band_amax = fp_max_arrays_band[index_amax]
+
+        result_dict = self.PolarizationFP(fp_band_data, fp_max_arrays_band_amax, band_amax)
+
+        # --- Add Incidence
+        fp_cf_waveforms = np.append(fp_cf_waveforms,
+                                    [result_dict["incidence"]],
+                                    axis=0)
+
+        # --- Add Modulus
+        fp_cf_waveforms = np.append(fp_cf_waveforms,
+                                    [result_dict["modulus"]],
+                                    axis=0)
+
+        # --- Bring everything to float32
+        fp_cf_waveforms = fp_cf_waveforms.astype("float32")
+
+        return fp_cf_waveforms
+
+    def get_stream(self):
+        return self.stream
+
+    def work(self, stream):
```

Comment:
From what I can see, this function assumes that the stream consists of exactly three traces, named `ZNE`, from a single station, and aligned in time. This is not given for streams passed to `annotate`. Instead, this work should be done in `annotate_batch_pre`.